src/wasm-binary.h

```diff
@@ -1367,6 +1369,21 @@ class WasmBinaryBuilder {
 
   std::vector<Expression*> expressionStack;
 
+  // Each let block in the binary adds new locals to the bottom of the index
+  // space. That is, all previously-existing indexes are bumped to higher
+  // indexes. getAbsoluteLocalIndex does this computation.
+  // Note that we must track not just the number of locals added in each let,
+  // but also their absolute indexes, as binaryen will add new locals as it goes
+  // for things like stacky code and tuples (so there isn't a simple way to get
+  // to the absolute index from a relative one). Hence each entry here is a list
+  // of the absolute indexes in use by that let.
```

Comment:
I think I'm missing something, but I still don't understand why we should track all indices by value. Even if locals added for tuples and stuff in between, when we start a new `let` scope, all `let` vars within that scope are bumped by a certain constant amount. What I mean is, if a `let` has two variables `a` and `b`, the amount bumped for `a` and `b` will be the same because they are in the same `let`. Then can't we keep just a single value per `let` and not all indices..? Please correct me if I'm missing something.

Reply:
That's correct, but it is more that previous lets are the issue. That is, lets can stack/nest, and we can add stacky locals in the middle. Consider this situation:

* We add a let for `$x` (just a made up name for this discussion - there is no name in the wasm binary, of course), giving it binaryen/absolute index 15.
* We add some locals for stacky code, using up indexes 16-30.
* We add a let for $y`, giving it binaryen/absolute index 31.
* We read a `local.get` with wasm/relative index 0. That's easy, it is `$y`, the last let that was "pushed". And it seems practical to compute 31 from a table of the pushed lets, I agree.
* We read a `local.get` with wasm/relative index 1. This is where it is harder. This refers to `$x`, but what is `$x`'s absolute index? We can't just subtract the lets that were pushed after it. If we just do that, we'd get 30, but that ignores the stacky locals. So we need to know about those.

And the messy part is that lets are stacked/nested, while stacky locals are not - they are added from there til the end. It is this mixture that makes it confusing to devise some sort of mechanism to track new locals. Instead, I think it's simpler to just remember the absolute indexes of all lets.

Reply:
I might be still mistaken, but I I thought is, rather than remembering absolute indices for all let vars, we can remember only two numbers (start offset and number of vars) per `let`. In your example, when we encounter the first `let` for `$x`, we push this pair to `letStack`:
```
letStack = [<15 (offset), 1 (# of vars in let)>]
```
And then we add locals for stacky code in 16-30.
And we encounter the second `let`, and we push one more entry to the `letStack`:
```
letStack = [<15 (offset), 1 (# of vars in let)>, <31 (offset), 1 (# of vars in let)>]
```

Now we read the binary index 0. Because it is within the range of the latest `let`, we can easily figure out it is the absolute index 31.
Then we read the binary index 1. Now it exceeds the size of the latest `let`, so we subtract the size of the latest `let` (1) from the index 1. Now we have the value 0. And we go to the second last entry, <15, 1>. And we know this binary index means the absolute index 15.

Reply:
Thanks! Yes, that is more efficient and no more complicated, so definitely a better solution. Done.